### PR TITLE
compiler/next: implement type construction

### PR DESCRIPTION
**Pull request by mppf**
This PR adds support for resolving type construction calls in the compiler/next prototype resolver.

This PR uses the approach of generating a TypedFnSignature for a type constructor and having that participate in candidate selection and disambiguation to allow other functions with the same name as a type. While we could decide to disallow this in the future, it seems to be a good idea to have the new resolver able to handle it.

For example, this program doesn't type resolve with the production compiler but the new resolver does handle it:

``` chapel
record R {
  type t;
}

{
  proc R(arg: int) {
    return arg;
  }

  var x = R(arg=1);
  var y:R(t=real);
}
```

This PR takes these main steps:
 * fills in the details for CompositeType to represent individual fields and adjusts the related classes appropriately.
 * makes some adjustments to the resolver and UntypedFnSignature and TypedFnSignature in particular in order to allow the compiler to add a TypedFnSignature during resolution that is not directly a function declared in the source code and parsed into the uAST. This strategy of creating a TypedFnSignature apart from a uAST Function declaration is also a strategy I am expecting to follow for compiler-generated default functions.
 * adds code to resolution-queries to resolve field types and to work with generated type constructors

This PR also includes some enabling changes:
* Fixes a some bugs in defaultUpdateOwned
* De-indents resolution-queries.h (to fit with practice of not indenting within namespaces and instead commenting on end of curly braces)
* Address public fields in UntypedFnSignature and TypedFnSignature and improve their API. Now they behave similarly to a Type with a `get` class method to create one in a unique-ified manner. Also, these types can now work with an ID that is not a Function - the ID can refer to say a Record declaration in uAST.
* Added a helper, `Type::getCompositeType` to return the receiver as a CompositeType if it is one, or, for a ClassType to return the referred to BasicClassType. (Why does ClassType refer to BasicClassType instead of just containing the fields? It is because we want to have just one type/instantiation for the fields of a class even as we can have various different management strategies used with it).
 * Adds a helper `PrimitiveType::getWithNameAndWidth` to make it easier to construct these types from a name and width.
 * Fixes a problem where TypeClassesList.h indicated ClassType is a CompositeType (it is not)
 * Adds some additional built-in types to `Type::gatherBuiltins`

Reviewed by @dlongnecke-cray - thanks!

### CODE REVIEW

**Review by dlongnecke-cray — APPROVED** [review]
Wow, this is a lot to digest. Truly awesome work! I'm going to approval now since the code looks good, but I'll probably read through another time.

File: compiler/next/lib/resolution/resolution-queries.cpp [hunk 1]
```diff
@@ -812,27 +1047,69 @@ const TypedFnSignature* instantiateSignature(Context* context,
   // compute the substitutions
   SubstitutionsMap substitutions;
   for (const FormalActual& entry : faMap.byFormalIdx) {
-    if (entry.formalType.isGenericOrUnknown()) {
+    // note: entry.actualType can have type()==nullptr and UNKNOWN.
+    // in that case, resolver code should treat it as a hint to
+    // use the default value. Unless the call used a ? argument.
+    if (call.hasQuestionArg &&
+        entry.actualType.kind() == QualifiedType::UNKNOWN &&
```
[the remark this answers is not in the document]
Comment (mppf) [reply compiler/next/lib/resolution/resolution-queries.cpp:1054]: We could add a helper for it, but today we have a difference between some things:
 * kind can be UNKNOWN
 * type can be `nullptr`
 * type can be `UnknownType`

I'm not really certain if we want to restrict the available combinations here. In particular, I have been thinking that `UNKNOWN`+`nullptr` indicates a lack of result or failure to compute a result, but perhaps `UNKNOWN`+`UnknownType` indicates to a user that the type cannot be known yet (e.g. in a recursive type construction).

Put another way, I haven't yet decided if we need to be able to use both `nullptr` and `UnknownType`.

**Comment by bradcray** [comment]
@mppf / @dlongnecke-cray : Reviewing this PR for a status update, I don't feel convinced that this program should resolve in Chapel, at least not without additional discussion:

```chapel
record R {
  type t;
}

{
  proc R(arg: int) {
    return arg;
  }

  var x = R(arg=1);
  var y:R(t=real);
}
```

Specifically, we don't generally support mixing different kinds of symbol overloads at a given scope, like:

```chapel
var x: int;
proc x { ... }
module x { ... }
type x = int;
```

and this program seems on the cusp of that.  On one hand, it's not putting the two R's at the same scope, but on the other, it seems to essentially be treating `record R` as though the only symbols it introduces into a scope are just additional function overloads that are not shadowed by an inner R, but just add to the overload set.  At first blush, this feels surprising to me and worthy of discussion where this PR seems to assert that it's behavior that we should've supported all along but didn't.

**Comment by dlongnecke-cray** [comment]
I'm not Michael, but I think in discussion he has recognized that we don't support this behavior today, but we'd like to engineer the capability to support the behavior in the future if we want.

**Comment by bradcray** [comment]
> he has recognized that we don't support this behavior today, but we'd like to engineer the capability to support the behavior in the future if we want.

I read:

> this program doesn't type resolve with the production compiler but the new resolver does handle it:

as meaning "The new compiler already does support it" (so "will" until/unless someone indicates that it shouldn't) rather than "It could be made to support it in the future if we choose to."  So that's why I was trying to indicate that I thought it potentially shouldn't and that the production compiler may be doing the right thing here already.  Or at least that it deserves discussion (as I feel like other potential changes or improvements in the handling of the language have when the new compiler has run into such cases).

**Comment by mppf** [comment]
@bradcray - that's a good point - I'll make an issue so we can discuss.

**Comment by bradcray** [comment]
I'll eventually add this to the issue, but capturing some thoughts:  Last night / this morning, I found myself wondering "would I object to this change, or was I just objecting to the lack of discussion on it?" and decided that I object to the change itself.  Specifically, I don't think that the following should be permitted

```chapel
record R { var x: int; }
{
  proc R() { ... }  // shadow record R

  proc foo(x: R) { ... }  // proc R shadows 'type R'
  type Ralias = R;  // ditto
  proc R.bar() { ... } // ditto
}
```

So while the `R(...)` expression is very similar to a function call in the type signature setting `var myR: R(...);` and the initializer call setting (`new R(...)`), I think that it should be linked to the type and not be "more visible" than the type name itself nor considered a candidate in the overload sets for `R` in that inner scope.

If we really wanted this, it suggests to me that we'd be going down the path of having different name categories for different symbol types (e.g. types are in one bucket, functions are in another), and that's not a change I'm excited about.  I prefer Chapel's relatively flat namespace across types, variables, procedures, modules, etc.

**Comment by mppf** [comment]
@bradcray - I've created #18817 and I didn't really copy your comment over but did try to summarize my understanding of it in the top of it. I'd like to encourage you to paste it in there as a comment or whatever makes sense.

I described in that issue why my reading of the language specification led me to believe that the type constructor should participate in overload resolution. I don't think you're necessarily saying here that shouldn't occur but it is certainly a related question.

**Comment by bradcray** [comment]
Thanks for opening the issue!  I transcribed and expanded upon my comment over there.  Let me know if I missed any key tendrils of the conversation that I should've addressed.  I gave it a pretty quick read due to feeling familiar with the core issue here.

**Pull request merged**